gax-java/gax-grpc/src/main/java/com/google/api/gax/grpc/GrpcLoggingInterceptor.java

```diff
@@ -0,0 +1,197 @@
+      public void start(Listener<RespT> responseListener, Metadata headers) {
+        logRequestInfo(method, logDataBuilder, LOGGER);
+        recordRequestHeaders(headers, logDataBuilder);
+        SimpleForwardingClientCallListener<RespT> responseLoggingListener =
+            new SimpleForwardingClientCallListener<RespT>(responseListener) {
+              @Override
+              public void onHeaders(Metadata headers) {
+                recordResponseHeaders(headers, logDataBuilder);
+                super.onHeaders(headers);
+              }
+
+              @Override
+              public void onMessage(RespT message) {
+                recordResponsePayload(message, logDataBuilder);
+                super.onMessage(message);
+              }
+
+              @Override
+              public void onClose(Status status, Metadata trailers) {
+                logResponse(status, logDataBuilder, LOGGER);
+                super.onClose(status, trailers);
+              }
+            };
+        currentListener = responseLoggingListener;
+        super.start(responseLoggingListener, headers);
+      }
+
+      @Override
+      public void sendMessage(ReqT message) {
+        logRequestDetails(message, logDataBuilder);
+        super.sendMessage(message);
+      }
+    };
+  }
+
+  // Helper methods for logging
+  // some duplications with http equivalent to avoid exposing as public method for now
+  <ReqT, RespT> void logRequestInfo(
+      MethodDescriptor<ReqT, RespT> method, LogData.Builder logDataBuilder, Logger logger) {
+    try {
+      if (logger.isInfoEnabled()) {
+        logDataBuilder.serviceName(method.getServiceName()).rpcName(method.getFullMethodName());
+
+        if (!logger.isDebugEnabled()) {
+          LoggingUtils.logWithMDC(
+              logger, Level.INFO, logDataBuilder.build().toMapRequest(), "Sending gRPC request");
+        }
+      }
+    } catch (Exception e) {
+      logger.error("Error logging request info (and headers)", e);
+    }
+  }
+
+  private void recordRequestHeaders(Metadata headers, LogData.Builder logDataBuilder) {
+    try {
+      if (LOGGER.isDebugEnabled()) {
+        JsonObject requestHeaders = mapHeadersToJsonObject(headers);
+        logDataBuilder.requestHeaders(GSON.toJson(requestHeaders));
+      }
+    } catch (Exception e) {
+      LOGGER.error("Error recording request headers", e);
+    }
+  }
+
+  void recordResponseHeaders(Metadata headers, LogData.Builder logDataBuilder) {
+    try {
+      if (LOGGER.isDebugEnabled()) {
+        JsonObject responseHeaders = mapHeadersToJsonObject(headers);
+        logDataBuilder.responseHeaders(GSON.toJson(responseHeaders));
+      }
+    } catch (Exception e) {
+      LOGGER.error("Error recording response headers", e);
+    }
+  }
+
+  <RespT> void recordResponsePayload(RespT message, LogData.Builder logDataBuilder) {
+    try {
+      if (LOGGER.isDebugEnabled()) {
+        logDataBuilder.responsePayload(GSON.toJsonTree(message));
```

Comment:
`message` is a Protobuf message, so we should use Protobuf utils to serialize it to Json if we have to. See how we serialized it in [ProtoRestSerializer](https://github.com/googleapis/sdk-platform-java/blob/main/gax-java/gax-httpjson/src/main/java/com/google/api/gax/httpjson/ProtoRestSerializer.java#L82). Same comment for the request payload logging below.

Reply:
Thanks for the pointer. Will look into.